File: ts/packages/core/src/palette-file.ts
import { z } from 'zod';
import { MAX_PALETTE, paletteEntryFrom } from './geometry/palette.js';
import { PaletteEntrySchema } from './geometry/schema.js';
import type { Palette } from './geometry/types.js';
import { err, ok, type Result } from './result.js';
import { resultFromZodError } from './zod-diagnostic.js';

// SPEC §6.10 (v0.7): external palette file — a shareable palette bound to
// a model via the manifest's top-level `palette` reference. The entry
// grammar, 62-color maximum and index assignment (`0-9a-zA-Z`) are the
// same as the inline geometry palette (§7.4); only the container differs.
// Object form (not a bare array) so the format has room for metadata
// (named colors etc.) without a breaking change.
//
// { "colors": ["#1a1a1a", { "color": "#f4c9a0", "metallic": 1 }, ...] }
//
// `colors` used to be `z.array(z.string())` with the hex check done by hand
// below. Sharing PaletteEntrySchema is what stops a palette FILE and an
// inline palette from disagreeing about what an entry may say — a shared
// palette that the geometry files could not express would be a poor kind
// of shared.

export const PaletteFileSchema = z
  .object({
    colors: z.array(PaletteEntrySchema).min(1).max(MAX_PALETTE),
  })
  .strict();

export function parsePaletteFile(json: unknown): Result<Palette> {
  const parsed = PaletteFileSchema.safeParse(json);
  if (!parsed.success) return resultFromZodError(parsed.error, json);
  const entries = [];
  for (const [i, doc] of parsed.data.colors.entries()) {
    const entry = paletteEntryFrom(doc);
    if (entry === null) {
      // Carries a path like every other reader's failures, so a caller
      // holding the text can locate it with `locateJsonPath`.
      const hex = typeof doc === 'string' ? doc : doc.color;
      return err('invalid-value', `colors.${i}: invalid color '${hex}'`, [
        'colors',
        i,
      ]);
    }
    entries.push(entry);
  }
  return ok(entries);
}

File: csharp/Cuboidy/PaletteFile.cs
// Port of ts/packages/core/src/palette-file.ts.
//
// SPEC §6.10 (v0.7): external palette file — a shareable palette bound to a
// model via a §7.4 reference or the manifest's top-level `palette`. The entry
// grammar, 64-color maximum and index assignment (`0-9a-zA-Z$%`) are the same as
// the inline geometry palette (§7.4); only the container differs. Object form
// (not a bare array) so the format has room for metadata — named colors and
// the like — without a breaking change.
//
//   { "colors": ["#1a1a1a", { "color": "#f4c9a0", "metallic": 1 }, …] }
//
// The entry reader is `GeometrySchema.ReadPaletteEntry`, shared rather than
// restated: a shared palette that the geometry files could not express would
// be a poor kind of shared.

using System;
using System.Collections.Generic;
using System.Text.Json;

namespace Cuboidy;

public static class PaletteFileReader
{
    public static Result<IReadOnlyList<PaletteEntry>> ParsePaletteFile(JsonElement json)
    {
        try
        {
            DocPath at = DocPath.Root;
            ObjectFields fields = JsonRead.Fields(json, at, "colors");

            DocPath colorsAt = at.Add("colors");
            JsonElement colorsEl = JsonRead.ArrayValue(fields.Required("colors"), colorsAt);
            int count = colorsEl.GetArrayLength();
            if (count < 1 || count > PaletteCodec.MaxPalette)
            {
                throw JsonRead.Fail(
                    CuboidyErrorCode.WrongArity,
                    $"expected 1..{PaletteCodec.MaxPalette} colors, got {count}",
                    colorsAt);
            }

            var entries = new List<PaletteEntryDoc>(count);
            int i = 0;
            foreach (JsonElement entryEl in colorsEl.EnumerateArray())
            {
                entries.Add(GeometrySchema.ReadPaletteEntry(entryEl, colorsAt.Add(i)));
                i++;
            }

            return Result.Ok(GeometryReader.ColorsToPalette(entries));
        }
        catch (ReaderException e)
        {
            return e.ToResult<IReadOnlyList<PaletteEntry>>();
        }
    }

    // The same thing from TEXT. Three callers in the reference had each wrapped
    // the reader in their own `JSON.parse` + try/catch, and all three swallowed
    // a malformed document into a bare `null` — indistinguishable from "no such
    // file" at the call site, so none of them could report what was actually
    // wrong.
    public static Result<IReadOnlyList<PaletteEntry>> ParsePaletteFileText(string text)
    {
        if (text is null) throw new ArgumentNullException(nameof(text));

        JsonDocument document;
        try
        {
            document = JsonDocument.Parse(text);
        }
        catch (JsonException e)
        {
            // Wording preserved from the reference, which preserved it in turn
            // from the reader it replaced — it is what the CLIs have always
            // printed, and a diagnostic's text is user-facing whether or not
            // anything parses it.
            return Result.Err<IReadOnlyList<PaletteEntry>>(
                CuboidyErrorCode.InvalidValue, $"JSON parse: {e.Message}");
        }

        using (document)
        {
            return ParsePaletteFile(document.RootElement);
        }
    }
}
